libraries/node-core-library/src/Resolve.ts

```diff
@@ -3,15 +3,23 @@
 
 import { sync } from 'resolve';
 import * as path from 'path';
-import { FileSystem, PackageJsonLookup } from '@rushstack/node-core-library';
+
+import { PackageJsonLookup } from './PackageJsonLookup';
+import { FileSystem } from './FileSystem';
 
 const packageJsonLookup: PackageJsonLookup = new PackageJsonLookup();
 
+/**
+ * @alpha
+ */
 export interface IResolvePackageOptions {
   doNotResolveSymlinks: boolean;
 }
 
-export class ResolveUtilities {
+/**
+ * @alpha
+ */
+export class Resolve {
   /**
    * Gets the path of a given package name from the perspective of a given path
    */
```

Comment:
Suggested API design:

```ts

interface IImportResolveOptions {
  packageName: string;  // required for now
  baseFolderPath: string; // required for now

  // If true, if the package name matches a Node.js system module, then the return
  // value will be the package name without any path.  This will take precedence over
  // an installed NPM package of the same name.
  //
  // Example:
  // Import.resolveModulePath({ packageName: "fs", baseFolderPath: process.cwd() })
  //  --> "fs"
  includeSystemModules?: boolean;

  // If true, then packageName is allowed to refer to the package.json of the active project.
  // It will take precedence over any installed dependency with the same name.
  // Note that this requires an additional PackageJsonLookup calculation.
  //
  // Example:
  // Import.resolveModulePath({ packageName: "my-project", baseFolderPath: process.cwd(), 
  //   allowSelfReference: true })
  //  --> "path/to/my-project"
  allowSelfReference?: boolean;
}

interface IImportResolveModulePathOptions extends IImportResolveOptions {
}
interface IImportResolvePackageFolderOptions extends IImportResolveOptions {
}

class Import {
  // Example:
  // Import.resolveModulePath({ packageName: "colors", baseFolderPath: process.cwd() })
  //  --> "path/to/node_modules/colors/lib/index.js"
  public static resolveModulePath(options: IImportResolveModulePathOptions): string {
  }

  // Example:
  // Import.resolvePackageFolder({ packageName: "colors", baseFolderPath: process.cwd() })
  //   --> "path/to/node_modules/colors"
  public static resolvePackageFolder(options: IResolvePackageFolderOptions): string {
  }
}

```

Reply:
Bonus points if you can figure out how to model the [realPathSync](https://github.com/microsoft/rushstack/blob/de6d949aa5cce8c9ed957b552c6e2fe3ba13dd0d/apps/rush-lib/src/logic/deploy/DeployManager.ts#L302) hook that `rush deploy` needs. :-)

And also resolution of module+path expressions (e.g. `"@rushstack/eslint-config/patch/modern-module-resolution"`)